File: 프로그래머스/2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.java
import java.util.*;
class Solution {
    LinkedList<Integer> brige = new LinkedList();
    LinkedList<Integer> que = new LinkedList();
    int sumWeight = 0;
    
    public int solution(int bridge_length, int weight, int[] truck_weights) {
        for(int i = 0; i < bridge_length; i++){
            brige.addLast(0);
        }
        
        for(int truck : truck_weights){
            que.addLast(truck);
        }
        
        int time = 0;
        
        while(!que.isEmpty() || sumWeight != 0){
            time += 1;
            int popItem = 0;
            sumWeight -= brige.pollFirst();
            if(!que.isEmpty() && sumWeight + que.peekFirst() <= weight){
                popItem = que.pollFirst();
                sumWeight += popItem;
                brige.addLast(popItem);
            } else{
                brige.addLast(0);
            }
        }
        return time;
    }

}

File: 프로그래머스/2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.js
function solution(bridge_length, weight, truck_weights) {
    let time = 0;
    let nowSum = 0;
    let nowLength = 0;
    const brige = Array(bridge_length).fill(0);
    while(truck_weights.length || nowLength){
        time += 1;
        const outTruck = brige.shift();
        nowSum -= outTruck;
        if(outTruck) {
            nowLength -= 1;
        }
        if(nowSum + truck_weights[0] <= weight && nowLength + 1 <= bridge_length){
            const truckWeight = truck_weights.shift();
            brige.push(truckWeight);
            nowSum += truckWeight;
            nowLength += 1
            continue
        }
        brige.push(0);
    }
    
    
    return time;
}